PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc

```diff
@@ -0,0 +1,161 @@
+                               int64_t batch_size) const {
+    // create input tensor objects from data values
+    std::vector<Value> input_tensors;
+    auto memory_info = MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
+    for (const auto& name : input_node_strings_) {
+      auto iter = std::find(input_names.begin(), input_names.end(), name);
+      if (iter == input_names.end()) {
+        throw cms::Exception("RuntimeError") << "Input " << name << " is not provided!";
+      }
+      auto value = input_values.begin() + (iter - input_names.begin());
+      auto input_dims = input_node_dims_.at(name);
+      input_dims[0] = batch_size;
+      auto input_tensor =
+          Value::CreateTensor<float>(memory_info, value->data(), value->size(), input_dims.data(), input_dims.size());
+      assert(input_tensor.IsTensor());
+      input_tensors.emplace_back(std::move(input_tensor));
+    }
+
+    // set output node names; will get all outputs if `output_names` is not provided
+    std::vector<const char*> run_output_node_names;
+    if (output_names.empty()) {
+      run_output_node_names = output_node_names_;
+    } else {
+      for (const auto& name : output_names) {
+        run_output_node_names.push_back(name.c_str());
+      }
+    }
+
+    // run
+    auto output_tensors = session_->Run(RunOptions{nullptr},
+                                        input_node_names_.data(),
+                                        input_tensors.data(),
+                                        input_tensors.size(),
+                                        run_output_node_names.data(),
+                                        run_output_node_names.size());
+
+    // convert output to floats
+    FloatArrays outputs;
+    for (auto& output_tensor : output_tensors) {
+      assert(output_tensor.IsTensor());
+
+      // get output shape
+      auto tensor_info = output_tensor.GetTensorTypeAndShapeInfo();
+      auto length = tensor_info.GetElementCount();
+
+      auto floatarr = output_tensor.GetTensorMutableData<float>();
+      outputs.emplace_back(floatarr, floatarr + length);
+    }
+    assert(outputs.size() == run_output_node_names.size());
+
+    return outputs;
+  }
+
+  const std::vector<std::string>& ONNXRuntime::getOutputNames() const {
+    if (session_) {
+      return output_node_strings_;
+    } else {
+      throw cms::Exception("RuntimeError") << "Needs to call createSession() first before getting the output names!";
+    }
+  }
+
+  const std::vector<int64_t>& ONNXRuntime::getOutputShape(const std::string& output_name) const {
+    auto iter = output_node_dims_.find(output_name);
+    if (iter == output_node_dims_.end()) {
+      throw cms::Exception("RuntimeError") << "Output name " << output_name << " is invalid!";
+    } else {
+      return iter->second;
+    }
+  }
+
+  Env& ONNXRuntime::getEnv() {
+    static Env env(ORT_LOGGING_LEVEL_WARNING, "onnxruntime");
```

Comment:
Can this be `static const`? Or is `Env` otherwise fully thread safe?

Reply:
It cannot be `const` due to: https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L171

> Or is Env otherwise fully thread safe?

As far as I can tell, it is the case.

Reply:
Could you add a comment briefly explaining that?